### PR TITLE
[wgpu-hal] Initial Vulkan ray query and acceleration structure support

### PR DESCRIPTION
**Pull request by expenses**
**Checklist**

- [ ] Run `cargo clippy`.
- [ ] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**

#1040 

**Description**

Still a work in progress, but I can ray-trace instanced triangles now:

20220912_23h26m53s_grim

Needs a cleanup and support for acceleration structure updates.

**Testing**
_Explain how this change is tested._

### CODE REVIEW

**Comment by expenses** [comment]
Besides documentation and dummy implementations for the other backends, I think this is ready for review.

File: wgpu-hal/src/vulkan/command.rs [hunk 1]
```diff
@@ -341,6 +341,115 @@ impl crate::CommandEncoder<super::Api> for super::CommandEncoder {
...
+                    ),
+            });
+
+        if desc.mode == crate::AccelerationStructureBuildMode::Update {
+            geometry_info.src_acceleration_structure = desc.destination_acceleration_structure.raw;
+        }
+
+        let geometry_info = geometry_info.build();
+
+        let range = &[range][..];
```
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/command.rs:444]: What on earth?

File: wgpu-hal/src/vulkan/device.rs [hunk 2]
```diff
@@ -788,7 +788,14 @@ impl crate::Device<super::Api> for super::Device {
...
                 size: req.size,
-                align_mask: req.alignment - 1,
+                align_mask: if desc
+                    .usage
+                    .contains(crate::BufferUses::TOP_LEVEL_ACCELERATION_STRUCTURE_INPUT)
+                {
+                    16
+                } else {
+                    req.alignment
+                } - 1,
```
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/device.rs:798]: Let's pull the if statement into a variable.

File: wgpu-hal/src/vulkan/device.rs [hunk 3]
```diff
@@ -1391,9 +1408,40 @@ impl crate::Device<super::Api> for super::Device {
...
+                        desc.acceleration_structures[start as usize..end as usize]
+                            .iter()
+                            .map(|acceleration_structure| acceleration_structure.raw),
+                    );
+
+                    let acceleration_structure_info =
+                        vk::WriteDescriptorSetAccelerationStructureKHR::builder()
+                            .acceleration_structures(&raw_acceleration_structures[raw_start..]);
+
+                    // todo: Dereference the struct to get around lifetime issues. Safe as long as we never resize
```
Comment (cwfitzgerald) [inline wgpu-hal/src/vulkan/device.rs:1428]: If we need to do this, let's add an assertion that we always have room on push.

File: wgpu-hal/examples/ray-traced-triangle/main.rs [hunk 4]
```diff
@@ -0,0 +1,934 @@
...
+                .unwrap()
+        };
+
+        let instances = [
+            Instance {
+                transform: transform_matrix,
+                instance_custom_index_and_mask: pack_24_8(0, 0xff),
+                instance_shader_binding_table_record_offset_and_flags: pack_24_8(0, 0),
+                acceleration_structure_reference: unsafe {
+                    device.get_acceleration_structure_device_address(&blas)
```
Comment (cwfitzgerald) [inline wgpu-hal/examples/ray-traced-triangle/main.rs:481]: This is the kind of thing we'd hide at the wgpu-core level?

File: wgpu-hal/src/lib.rs [hunk 5]
```diff
@@ -660,6 +684,9 @@ bitflags::bitflags! {
         const STORAGE_READ_WRITE = 1 << 8;
         /// The indirect or count buffer in a indirect draw or dispatch.
         const INDIRECT = 1 << 9;
+        const ACCELERATION_STRUCTURE_SCRATCH = 1 << 10;
+        const BOTTOM_LEVEL_ACCELERATION_STRUCTURE_INPUT = 1 << 11;
+        const TOP_LEVEL_ACCELERATION_STRUCTURE_INPUT = 1 << 12;
```
Comment (cwfitzgerald) [inline wgpu-hal/src/lib.rs:689]: These probably need to be added to either INCLUSIVE or EXCLUSIVE in some way?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
This is really great work - have no massive comments. tbh I don't fully understand RT apis as I haven't used them, but from what I know about them, this design seems sane and low overhead.

**Comment by expenses** [comment]
This PR is currently on-pause because I don't have the time to work on it along with everything else I'm doing. I don't think it'll take much to get it to a mergeable state and if someone else wants to work on that then I'm happy to review their changes.

**Comment by seabassjh** [comment]
@cwfitzgerald is there a list of TODOs left on this before being mergeable?  I'd like to pick this up

**Comment by cwfitzgerald** [comment]
@seabassjh love to have this be picked up - the main ones are the comments on the PR

**Comment by expenses** [comment]
> @cwfitzgerald is there a list of TODOs left on this before being mergeable? I'd like to pick this up

I imagine there's some work to be done to resolve https://github.com/gfx-rs/wgpu/pull/3044 etc as well

**Comment by daniel-keitel** [comment]
I would like to help out.
@seabassjh have you continued work on this PR?
I have already made most of the requested changes, but I don't know how to contribute since it is @expenses PR.

**Comment by expenses** [comment]
@daniel-keitel if you link me your changes I can review them and merge them into this branch

**Comment by daniel-keitel** [comment]
@expenses I created a pull request on your fork.

**Comment by expenses** [comment]
Hi @daniel-keitel, sorry I missed this. It might be better if you created a new PR on wgpu and then the discussion moved over to there.

**Comment by cwfitzgerald** [comment]
Closing on favor of 3507

**Pull request closed without merging**